astropy/units/quantity.py

```diff
@@ -383,10 +383,9 @@ def __array_prepare__(self, obj, context=None):
             # decomposed, which involves being scaled by a float, but since
             # the array is an integer the output then gets converted to an int
             # and truncated.
-            result_dtype = np.result_type(*(args + tuple(
-                (float if converter and converter(1.) % 1. != 0. else int)
-                for converter in converters)))
-            if not np.can_cast(result_dtype, obj.dtype):
+            result_dtype = np.result_type(*((args +(float,)) if any(converters)
```

Comment:
Stupid nitpick about missing space after `+`.

Makes sense otherwise--nice that this is much cleaner.